designs/0005-static-matrices.md

```diff
@@ -0,0 +1,240 @@
+- Feature Name: static_matrices
+- Start Date: April 30th, 2020
+- RFC PR: [Example Branch](https://github.com/stan-dev/math/compare/feature/var-template)
+- Stan Issue: [#1805](https://github.com/stan-dev/math/issues/1805)
+
+# Summary
+[summary]: #summary
+
+This proposes a generalization of the reverse mode autodiff type in Stan Math,
+`var_value<T>`, where `T` is the underlying storage type of the autodiff variable.
+For `T == double`, this type is equivalent to the current Stan Math `var` type.
+
+The usefulness of this type comes from setting `T` to more general data structures,
+in particular the Eigen Matrix types. Instead of dealing with matrices of vars,
+this allows operations in the Math library to be written in terms of vars that are
+themselves matrices.
+
+This makes working with matrix variables in Stan autodiff much faster. This design
+is fully backwards compatible with the curren Stan autodiff types and those remain.
+
+For the rest of this proposal, "autodiff" means "reverse mode autodiff".
+
+With optimizations from the compiler the conditions for a `matrix` to be a
+constant matrix can be detected and applied to Stan programs automatically
+for users.
+
+# Motivation
+[motivation]: #motivation
+
+Currently, an `N x M` matrix in Stan is represented as an Eigen matrix holding
+a pointer to an underlying array of autodiff objects, each of those holding
+a pointer to the underlying autodiff object implementation.
+
+The basic Eigen type used by Stan is `Eigen::Matrix<var, R, C>`. For brevity,
+this will be referred to as a "matvar" type (a matrix of vars). There are no
```

Comment:
I think it'd help for readability to just use code styling for both `matvar` and `varmat`. It's done below in a couple spots and for consistency and being able to spot it in text, I think it might be useful.

Reply:
Yeah, I will do this.

I originally wanted to avoid `varmat` because `varmat` is not real code and `var` s are and I don't highlight them.